### PR TITLE
Flatten inheritance for DDPSharded strategy

### PR DESCRIPTION
**Pull request by edward-io**
## What does this PR do?

DDPShardedStrategy now inherits from ParallelStrategy instead of DDPStrategy.

Part of #https://github.com/PyTorchLightning/pytorch-lightning/issues/11863

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [ ] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/strategies/sharded.py [hunk 1]
```diff
@@ -11,46 +11,168 @@
...
+    def root_device(self) -> torch.device:
+        return self.parallel_devices[self.local_rank]
+
+    @num_nodes.setter
+    def num_nodes(self, num_nodes: int) -> None:
+        # note that world ranks is related to num_nodes, when resetting it, need to reset world ranks
+        self._num_nodes = num_nodes
+
+    @property
+    def num_processes(self):
```
Comment (ananthsub) [inline pytorch_lightning/strategies/sharded.py:96]:
```suggestion
    def num_processes(self) -> int:
```

File: pytorch_lightning/strategies/sharded.py [hunk 2]
```diff
@@ -11,46 +11,168 @@
...
+    @property
+    def distributed_sampler_kwargs(self):
+        distributed_sampler_kwargs = dict(num_replicas=len(self.parallel_devices), rank=self.global_rank)
+        return distributed_sampler_kwargs
+
+    def setup_environment(self) -> None:
+        self.setup_distributed()
+        super().setup_environment()
+
+    def setup_distributed(self):
```
Comment (ananthsub) [inline pytorch_lightning/strategies/sharded.py:108]: make this a protected function as it's not required by the Strategy API. 
Alternatively, its implementation can be folded into `setup_environment`

File: pytorch_lightning/strategies/sharded.py [hunk 3]
```diff
@@ -11,46 +11,168 @@
...
+        log.detail(f"{self.__class__.__name__}: setting up distributed...")
+        reset_seed()
+
+        # determine which process we are and world size
+        self.set_world_ranks()
+
+        # set warning rank
+        rank_zero_only.rank = self.global_rank
```
Comment (ananthsub) [inline pytorch_lightning/strategies/sharded.py:116]:
```suggestion
```
this line is redundant due to L128

File: pytorch_lightning/strategies/sharded.py [hunk 4]
```diff
@@ -11,46 +11,168 @@
...
+
+        # set up server using proc 0's ip address
+        # try to init for 20 times at max in case ports are taken
+        # where to store ip_table
```
Comment (ananthsub) [inline pytorch_lightning/strategies/sharded.py:120]:
```suggestion
```
comment is outdated

File: pytorch_lightning/strategies/sharded.py [hunk 5]
```diff
@@ -11,46 +11,168 @@
...
+        init_dist_connection(self.cluster_environment, self.torch_distributed_backend)
+
+    def set_world_ranks(self) -> None:
+        if self.cluster_environment is None:
+            return
+        self.cluster_environment.set_global_rank(self.node_rank * self.num_processes + self.local_rank)
+        self.cluster_environment.set_world_size(self.num_nodes * self.num_processes)
+        rank_zero_only.rank = self.cluster_environment.global_rank()
+
     def configure_ddp(self) -> None:
```
Comment (ananthsub) [inline pytorch_lightning/strategies/sharded.py:130]: 1. configure_ddp isn't called on its own. this is currently being called by DDP's `setup` method. Override the `setup` method on the strategy base class here for this to be called
2. this can be a private method, since it's not required by the strategy API

File: pytorch_lightning/strategies/sharded.py [hunk 6]
```diff
@@ -11,46 +11,168 @@
...
+
+        # set warning rank
+        rank_zero_only.rank = self.global_rank
+
+        # set up server using proc 0's ip address
+        # try to init for 20 times at max in case ports are taken
+        # where to store ip_table
+        init_dist_connection(self.cluster_environment, self.torch_distributed_backend)
+
+    def set_world_ranks(self) -> None:
```
Comment (ananthsub) [inline pytorch_lightning/strategies/sharded.py:123]: same here, let's make this protected

File: pytorch_lightning/strategies/sharded.py [hunk 7]
```diff
@@ -140,3 +259,49 @@ def register_strategies(cls, strategy_registry: Dict) -> None:
             cls,
             description=f"{cls.__class__.__name__}",
         )
+
+    def reconciliate_processes(self, trace: str) -> None:
+        if self.world_size < 2:
+            return
+
+        if not self._should_run_deadlock_detection():
```
Comment (ananthsub) [inline pytorch_lightning/strategies/sharded.py:267]: where is this defined? this is defined in DDP here: https://github.com/PyTorchLightning/pytorch-lightning/blob/cf64f3443474a93d23b5afb0417e4a60298006e6/pytorch_lightning/strategies/ddp.py#L352-L359

**Review by ananthsub — CHANGES_REQUESTED** [review]

File: pytorch_lightning/strategies/sharded.py [hunk 8]
```diff
@@ -11,46 +11,201 @@
...
+        init_dist_connection(self.cluster_environment, self.torch_distributed_backend)
+
+    def _set_world_ranks(self) -> None:
+        if self.cluster_environment is None:
+            return
+        self.cluster_environment.set_global_rank(self.node_rank * self.num_processes + self.local_rank)
+        self.cluster_environment.set_world_size(self.num_nodes * self.num_processes)
+        rank_zero_only.rank = self.cluster_environment.global_rank()
+
+    def _configure_ddp(self) -> None:
```
Comment (ananthsub) [inline pytorch_lightning/strategies/sharded.py:145]: let's not carry over the DDP name here since this is a different module wrapper.

additionally, this can receive the trainer as an argument to avoid L146: self.lightning_module.trainer can simply be trainer
```suggestion
    def _configure_sdp(self, trainer: "pl.Trainer") -> None:
```

File: pytorch_lightning/strategies/sharded.py [hunk 9]
```diff
@@ -11,46 +11,201 @@
...
         trainer = self.lightning_module.trainer
-        if "reduce_buffer_size" not in self._ddp_kwargs:
+        if "reduce_buffer_size" not in self._kwargs:
             # For multi-node training, enabling bucketing will improve performance.
-            self._ddp_kwargs["reduce_buffer_size"] = self._REDUCE_BUFFER_SIZE_DEFAULT if self.num_nodes > 1 else 0
+            self._kwargs["reduce_buffer_size"] = self._REDUCE_BUFFER_SIZE_DEFAULT if self.num_nodes > 1 else 0
 
         self.model, self.optimizers = self._setup_model_and_optimizers(
             model=LightningShardedDataParallel(self.model),
             optimizers=trainer.optimizers,
```
Comment (ananthsub) [inline pytorch_lightning/strategies/sharded.py:153]: we need to rework this now that the strategy owns the optimizers. we can do that in a separate PR

File: pytorch_lightning/strategies/sharded.py [hunk 10]
```diff
@@ -11,46 +11,201 @@
...
+    def broadcast(self, obj: object, src: int = 0) -> object:
+        obj = [obj]
+        if self.global_rank != src:
+            obj = [None]
+        torch.distributed.broadcast_object_list(obj, src, group=_group.WORLD)
+        return obj[0]
+
+    def model_to_device(self) -> None:
+        log.detail(f"{self.__class__.__name__}: moving model to device [{self.root_device}]...")
+        self.model.to(self.root_device)
```
Comment (ananthsub) [inline pytorch_lightning/strategies/sharded.py:173]: super nit: let's move this up in the file to be right under `setup` so that all the collective implementations are in a row

File: pytorch_lightning/strategies/sharded.py [hunk 11]
```diff
@@ -11,46 +11,201 @@
...
+    def validation_step(self, *args, **kwargs) -> Optional[STEP_OUTPUT]:
+        with self.precision_plugin.val_step_context():
+            if isinstance(self.model, ShardedDataParallel):
+                # used when calling `trainer.fit`
+                return self.model(*args, **kwargs)
+            else:
+                # used when calling `trainer.validate`
+                return self.lightning_module.validation_step(*args, **kwargs)
+
+    def _determine_ddp_device_ids(self):
```
Comment (ananthsub) [inline pytorch_lightning/strategies/sharded.py:204]: same as above, let's drop the ddp terminology
```suggestion
    def _determine_device_ids(self):
```

if this isn't used outside of `barrier` above, we can also move this directly to be within the `barrier` definition

File: tests/strategies/test_sharded_strategy.py [hunk 12]
```diff
@@ -218,7 +218,7 @@ def on_predict_start(self) -> None:
 
 
 @RunIf(skip_windows=True, fairscale=True)
-def test_configure_ddp(tmpdir):
+def test__configure_ddp(tmpdir):
```
Comment (ananthsub) [inline tests/strategies/test_sharded_strategy.py:221]:
```suggestion
def test_configure_sdp(tmpdir):
```

File: tests/strategies/test_sharded_strategy.py [hunk 13]
```diff
@@ -232,13 +232,31 @@ def test_configure_ddp(tmpdir):
...
+        strategy._configure_ddp()
+    args, kwargs = mock_sharded.call_args
+    assert "reduce_fp16" in kwargs
+    assert kwargs["reduce_fp16"]
+
+
+# TODO (edward-io): merge test_custom_kwargs_sharded and test_custom_kwargs_sharded_spawn into one test after sharded_spawn is refactored
+@RunIf(skip_windows=True, fairscale=True)
+@mock.patch("pytorch_lightning.strategies.DDPShardedStrategy._wrap_optimizers", autospec=True)
+@pytest.mark.parametrize("cls", [DDPSpawnShardedStrategy])
```
Comment (ananthsub) [inline tests/strategies/test_sharded_strategy.py:253]: same here

**Review by ananthsub — COMMENTED** [review]
Nice! This will make simplifying optimizer creation & ownership easier in follow ups

File: pytorch_lightning/strategies/sharded.py [hunk 14]
```diff
@@ -11,46 +11,201 @@
...
+    def num_nodes(self, num_nodes: int) -> None:
+        # note that world ranks is related to num_nodes, when resetting it, need to reset world ranks
+        self._num_nodes = num_nodes
+
+    @property
+    def num_processes(self) -> int:
+        return len(self.parallel_devices) if self.parallel_devices is not None else 0
+
+    @property
+    def distributed_sampler_kwargs(self) -> dict[str, int]:
```
Comment (ananthsub) [inline pytorch_lightning/strategies/sharded.py:103]:
```suggestion
    def distributed_sampler_kwargs(self) -> Dict[str, int]:
```

File: pytorch_lightning/strategies/sharded.py [hunk 15]
```diff
@@ -11,46 +11,201 @@
...
+            accelerator=accelerator,
+            parallel_devices=parallel_devices,
+            cluster_environment=cluster_environment,
+            checkpoint_io=checkpoint_io,
+            precision_plugin=precision_plugin,
+        )
+        self._num_nodes: int = 1
+        self._rank_0_will_call_children_scripts: bool = False
+        self.sync_batchnorm: bool = False
+        self._kwargs = kwargs
```
Comment (ananthsub) [inline pytorch_lightning/strategies/sharded.py:83]:
```suggestion
        self._kwargs = kwargs
        self._set_world_ranks()
```

File: pytorch_lightning/strategies/sharded.py (hunk 6 above)
Comment (edward-io) [reply pytorch_lightning/strategies/sharded.py]: This is needed in accelerator_connector:

https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/trainer/connectors/accelerator_connector.py#L738-L739

We can revisit this in #7650

**Review by ananthsub — APPROVED** [review]

File: pytorch_lightning/strategies/sharded.py [hunk 16]
```diff
@@ -124,9 +276,6 @@ def block_backward_sync(self) -> Generator:
         else:
             yield None
 
-    def post_training_step(self):
```
Comment (ananthsub) [inline pytorch_lightning/strategies/sharded.py:127]: the `pre_backward` override above can be removed now

**Comment by edward-io** [comment]
I'm not sure why the TPU tests are failing here :(

**Comment by edward-io** [comment]
rebase on top of #11754

File: pytorch_lightning/strategies/sharded.py [hunk 17]
```diff
@@ -11,46 +11,197 @@
...
+        self.model.to(self.root_device)
+
+    def set_world_ranks(self) -> None:
+        if self.cluster_environment is None:
+            return
+        self.cluster_environment.set_global_rank(self.node_rank * self.num_processes + self.local_rank)
+        self.cluster_environment.set_world_size(self.num_nodes * self.num_processes)
+        rank_zero_only.rank = self.cluster_environment.global_rank()
+
+    def _configure_sdp(self, trainer: "pl.Trainer") -> None:
```
Comment (four4fish) [inline pytorch_lightning/strategies/sharded.py:146]: Should we align all of the _configure_x function name?

File: pytorch_lightning/strategies/sharded.py [hunk 18]
```diff
@@ -11,46 +11,197 @@
...
+            accelerator=accelerator,
+            parallel_devices=parallel_devices,
+            cluster_environment=cluster_environment,
+            checkpoint_io=checkpoint_io,
+            precision_plugin=precision_plugin,
+        )
+        self._num_nodes: int = 1
+        self._rank_0_will_call_children_scripts: bool = False
+        self._kwargs = kwargs
+        self.set_world_ranks()
```
Comment (tchaton) [inline pytorch_lightning/strategies/sharded.py:83]: What is it called there ?

File: pytorch_lightning/strategies/sharded.py [hunk 19]
```diff
@@ -11,46 +11,197 @@
...
+
+    @property
+    def num_nodes(self) -> int:
+        return self._num_nodes
+
+    @property
+    def root_device(self) -> torch.device:
```
Comment (tchaton) [inline pytorch_lightning/strategies/sharded.py:90]: Where is the `is_distributed` property. Without it, DistriburedSampler isn't injected in the DataLoader.

File: pytorch_lightning/strategies/sharded.py [hunk 20]
```diff
@@ -11,46 +11,197 @@
...
+    @num_nodes.setter
+    def num_nodes(self, num_nodes: int) -> None:
+        # note that world ranks is related to num_nodes, when resetting it, need to reset world ranks
+        self._num_nodes = num_nodes
+
+    @property
+    def num_processes(self) -> int:
+        return len(self.parallel_devices) if self.parallel_devices is not None else 0
+
+    @property
```
Comment (tchaton) [inline pytorch_lightning/strategies/sharded.py:102]: Careful, this doesn't match master.

File: pytorch_lightning/strategies/sharded.py [hunk 21]
```diff
@@ -11,46 +11,197 @@
...
+        log.detail(f"{self.__class__.__name__}: setting up distributed...")
+        reset_seed()
+
+        # determine which process we are and world size
+        self.set_world_ranks()
+
+        init_dist_connection(self.cluster_environment, self.torch_distributed_backend)
+        super().setup_environment()
+
+    def setup(self, trainer: "pl.Trainer") -> None:
```
Comment (tchaton) [inline pytorch_lightning/strategies/sharded.py:117]: I am definitely not a fan of this ...

File: pytorch_lightning/strategies/sharded.py (hunk 17 above)
Comment (tchaton) [reply pytorch_lightning/strategies/sharded.py:146]: Yes. This is still a `configure_ddp`. Consistency is key for any new contributors.

File: pytorch_lightning/strategies/sharded.py [hunk 22]
```diff
@@ -11,46 +11,197 @@
...
+
+    def set_world_ranks(self) -> None:
+        if self.cluster_environment is None:
+            return
+        self.cluster_environment.set_global_rank(self.node_rank * self.num_processes + self.local_rank)
+        self.cluster_environment.set_world_size(self.num_nodes * self.num_processes)
+        rank_zero_only.rank = self.cluster_environment.global_rank()
+
+    def _configure_sdp(self, trainer: "pl.Trainer") -> None:
+        if "reduce_buffer_size" not in self._kwargs:
```
Comment (tchaton) [inline pytorch_lightning/strategies/sharded.py:147]: Why this change?

File: pytorch_lightning/strategies/sharded.py [hunk 23]
```diff
@@ -59,7 +210,7 @@ def _setup_model_and_optimizers(self, model: Module, optimizers: List[Optimizer]
             and a list of optimizer wrapped in :class:~`fairscale.optim.OSS`.
         """
         optimizers = self._wrap_optimizers(optimizers)
-        model = ShardedDataParallel(model, sharded_optimizer=optimizers, **self._ddp_kwargs)
+        model = ShardedDataParallel(model, sharded_optimizer=optimizers, **self._kwargs)
```
Comment (tchaton) [inline pytorch_lightning/strategies/sharded.py:213]: Let's keep it as it was.

**Review by tchaton — COMMENTED** [review]
I really don't know what to think of this work.

We are moving from 143 lines to 345 lines. What is happening to PyTorch Lightning creating abstraction to make complex concepts easier to read and understand...
